File: baml_language/sdk_tests/crates/rust/type_shapes/customizable/test_complex_models.rs
//! Roundtrip coverage for a complex nested object graph.

// Anonymous unions surface as one synthesized enum per (null-stripped)
// union shape, named by joining the arm names with `Or` — literal-string
// arms as unit variants from their value, class/primitive arms wrapping
// their payload — and a trailing `| null` lowering to `Option<...>`
// around the enum.
use baml_bridge::{
    Map, baml_value::internal::__BamlValuePrivate, wire::inbound_value::Value as WireValue,
};
use baml_sdk::complex_models::{
    AccountTier, AuditEvent, CardPayment, CardPaymentOrWirePayment, ComplexProfile, ContactMethod,
    CreatedOrUpdatedOrApproved, DraftOrSentOrPaid, GeoPoint, GraphQueryOrGraphDiff,
    IntOrStringOrBool, Invoice, InvoiceOrPostalAddressOrString, LineItem,
    LiteralUnionIdentifierEdges, PostalAddress, ProfileOwner, Utf8LossyOrUtf8Strict, WirePayment,
    round_trip_complex_profile, round_trip_literal_union_identifier_edges,
};

// SDK_PARITY_LINT(skip): exercises Rust identifier synthesis for string-literal union arms
#[test]
fn test_complex_models_round_trip_non_identifier_string_literal_union_arms() {
    for (command, command_wire_value) in [
        (GraphQueryOrGraphDiff::GraphQuery, "graph.query"),
        (GraphQueryOrGraphDiff::GraphDiff, "graph.diff"),
    ] {
        assert_string_wire_value(&command, command_wire_value);
        for (encoding, encoding_wire_value) in [
            (Utf8LossyOrUtf8Strict::Utf8Lossy, "utf8-lossy"),
            (Utf8LossyOrUtf8Strict::Utf8Strict, "utf8-strict"),
        ] {
            assert_string_wire_value(&encoding, encoding_wire_value);
            let value = LiteralUnionIdentifierEdges {
                command: command.clone(),
                encoding,
            };
            assert_eq!(
                round_trip_literal_union_identifier_edges(value.clone()).unwrap(),
                value
            );
        }
    }
}

fn assert_string_wire_value<T: __BamlValuePrivate>(value: &T, expected: &str) {
    match value.to_baml().value {
        Some(WireValue::StringValue(actual)) => assert_eq!(actual, expected),
        other => panic!("expected string wire value {expected:?}, got {other:?}"),
    }
}

#[test]
fn test_complex_models_round_trip_complex_profile_preserves_deeply_nested_mixed_shape_class() {
    let home = PostalAddress {
        line1: "1 Compiler Way".to_string(),
        line2: None,
        city: "San Francisco".to_string(),
        region: "CA".to_string(),
        postal_code: "94107".to_string(),
        location: Some(GeoPoint {
            lat: 37.7749,
            lng: -122.4194,
        }),
    };
    let office = PostalAddress {
        line1: "200 Type Lane".to_string(),
        line2: Some("Suite 42".to_string()),
        city: "Oakland".to_string(),
        region: "CA".to_string(),
        postal_code: "94612".to_string(),
        location: None,
    };
    let email = ContactMethod {
        label: "email".to_string(),
        value: "ada@example.com".to_string(),
        verified: true,
    };
    let phone = ContactMethod {
        label: "phone".to_string(),
        value: "+1-555-0100".to_string(),
        verified: false,
    };
    let invoice = Invoice {
        id: "inv-001".to_string(),
        status: DraftOrSentOrPaid::Sent,
        items: vec![LineItem {
            sku: "sdk-pro".to_string(),
            quantity: 2,
            unit_price: 19.5,
            tags: vec!["sdk".to_string(), "typescript".to_string()],
            attributes: Map::from([
                ("language".to_string(), "ts".to_string()),
                ("support".to_string(), "priority".to_string()),
            ]),
        }],
        payment: Some(CardPaymentOrWirePayment::CardPayment(CardPayment {
            brand: "visa".to_string(),
            last4: "4242".to_string(),
            billing_address: home.clone(),
        })),
        notes: Some("first invoice".to_string()),
    };
    let wire_invoice = Invoice {
        id: "inv-002".to_string(),
        status: DraftOrSentOrPaid::Paid,
        items: vec![LineItem {
            sku: "sdk-enterprise".to_string(),
            quantity: 1,
            // DIVERGENCE(rust): python passes the int literal 250 and relies
            // on pydantic's construction-time int→float coercion; Rust writes
            // the float value directly.
            unit_price: 250.0,
            tags: vec!["sdk".to_string(), "enterprise".to_string()],
            attributes: Map::from([
                ("language".to_string(), "python".to_string()),
                ("term".to_string(), "annual".to_string()),
            ]),
        }],
        payment: Some(CardPaymentOrWirePayment::WirePayment(WirePayment {
            bank_name: "Boundary Bank".to_string(),
            routing_code: "110000000".to_string(),
            reference: None,
        })),
        notes: None,
    };
    let profile = ComplexProfile {
        id: "profile-001".to_string(),
        tier: AccountTier::Enterprise,
        owner: ProfileOwner {
            name: "Ada Lovelace".to_string(),
            primary_contact: email,
            backup_contacts: vec![phone],
        },
        addresses: vec![home, office],
        invoices: vec![invoice.clone(), wire_invoice],
        audit_trail: vec![
            AuditEvent {
                actor: "system".to_string(),
                action: CreatedOrUpdatedOrApproved::Created,
                context: Map::from([("source".to_string(), "fixture".to_string())]),
            },
            AuditEvent {
                actor: "reviewer".to_string(),
                action: CreatedOrUpdatedOrApproved::Approved,
                context: Map::from([
                    ("level".to_string(), "2".to_string()),
                    ("region".to_string(), "us".to_string()),
                ]),
            },
        ],
        metadata: Map::from([
            ("cohort".to_string(), "beta".to_string()),
            ("owner_kind".to_string(), "internal".to_string()),
        ]),
        featured: Some(InvoiceOrPostalAddressOrString::Invoice(invoice)),
        flags: vec![
            IntOrStringOrBool::Int(7),
            IntOrStringOrBool::String("manual-review".to_string()),
            IntOrStringOrBool::Bool(true),
        ],
    };

    assert_eq!(
        round_trip_complex_profile(profile.clone()).unwrap(),
        profile
    );
}

File: baml_language/sdk_tests/crates/swift/type_shapes/customizable/TestComplexModels.swift
// Roundtrip coverage for a complex nested object graph — port of
// python_pydantic2 `test_complex_models.py`, on the BamlUnionN family.
//
// Union-typed slots take positional cases (`payment: .t0(card)`);
// literal-union fields (`status`, `action`) collapse to String (no
// generated literal enums — the engine validates values).
import XCTest
import Baml
import BamlBridge

final class TestComplexModels: XCTestCase {
    func test_complex_models_round_trip_complex_profile_preserves_deeply_nested_mixed_shape_class() throws {
        typealias M = Baml.complex_models

        let home = M.PostalAddress(
            line1: "1 Compiler Way",
            line2: nil,
            city: "San Francisco",
            region: "CA",
            postal_code: "94107",
            location: M.GeoPoint(lat: 37.7749, lng: -122.4194)
        )
        let office = M.PostalAddress(
            line1: "200 Type Lane",
            line2: "Suite 42",
            city: "Oakland",
            region: "CA",
            postal_code: "94612",
            location: nil
        )
        let email = M.ContactMethod(label: "email", value: "ada@example.com", verified: true)
        let phone = M.ContactMethod(label: "phone", value: "+1-555-0100", verified: false)
        let invoice = M.Invoice(
            id: "inv-001",
            status: "sent",
            items: [
                M.LineItem(
                    sku: "sdk-pro",
                    quantity: 2,
                    unit_price: 19.5,
                    tags: ["sdk", "typescript"],
                    attributes: ["language": "ts", "support": "priority"]
                )
            ],
            payment: .t0(M.CardPayment(brand: "visa", last4: "4242", billing_address: home)),
            notes: "first invoice"
        )
        let wireInvoice = M.Invoice(
            id: "inv-002",
            status: "paid",
            items: [
                M.LineItem(
                    sku: "sdk-enterprise",
                    quantity: 1,
                    unit_price: 250,
                    tags: ["sdk", "enterprise"],
                    attributes: ["language": "python", "term": "annual"]
                )
            ],
            payment: .t1(
                M.WirePayment(
                    bank_name: "Boundary Bank",
                    routing_code: "110000000",
                    reference: nil
                )
            ),
            notes: nil
        )
        let profile = M.ComplexProfile(
            id: "profile-001",
            tier: .Enterprise,
            owner: M.ProfileOwner(
                name: "Ada Lovelace",
                primary_contact: email,
                backup_contacts: [phone]
            ),
            addresses: [home, office],
            invoices: [invoice, wireInvoice],
            audit_trail: [
                M.AuditEvent(actor: "system", action: "created", context: ["source": "fixture"]),
                M.AuditEvent(
                    actor: "reviewer",
                    action: "approved",
                    context: ["level": "2", "region": "us"]
                ),
            ],
            metadata: ["cohort": "beta", "owner_kind": "internal"],
            featured: .t0(invoice),
            flags: [.t0(7), .t1("manual-review"), .t2(true)]
        )

        let result = try M.round_trip_complex_profile(profile: profile)
        XCTAssertEqual(result, profile)

        // Class-arm identity survives the round trip: the wire says
        // WHICH payment class was selected, decode doesn't guess.
        XCTAssertNotNil(result.invoices[0].payment?.t0)
        XCTAssertNotNil(result.invoices[1].payment?.t1)
    }
}
